File: app/src/main/java/artwork/java/GameView.java
package artwork.java;

import android.content.Context;

import android.graphics.*;
import android.util.AttributeSet;
import android.view.MotionEvent;
import artwork.kotlin.ProgressListener;
import artwork.kotlin.UpdatePathsListener;

import java.util.Vector;

public class GameView extends androidx.appcompat.widget.AppCompatImageView {
    public static Canvas canvas;
    public RectF rect = new RectF();
    public Bitmap bitmap;
    public Paint paintBmp;
    public Paint paint;
    public Paint clearPaint;
    public Path path;
    public float sx;
    public float sy;
    public int penSize = 10;
    public Vector<Path> paths = new Vector<>();
    public int currentPathId = 0;
    boolean needRedraw = true;
    ProgressListener progressListener;
    UpdatePathsListener pathsListener;
    BrushManager brushManager = new BrushManager();
    Task task = null;
    /**
     * @param context the context
     */
    public GameView(Context context) {
        super(context);
    }
    public GameView(Context context, AttributeSet attrs) {
        super(context, attrs);
        progressListener = (ProgressListener)context;
        pathsListener = (UpdatePathsListener)context;
        path = new Path();
        paint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paint.setAntiAlias(true);
        paint.setStyle(Paint.Style.STROKE);
        paint.setStrokeJoin(Paint.Join.ROUND);
        paint.setStrokeCap(Paint.Cap.ROUND);
        paint.setPathEffect(new CornerPathEffect(10.0f));
        clearPaint = new Paint(Paint.ANTI_ALIAS_FLAG);
        paintBmp = new Paint(Paint.ANTI_ALIAS_FLAG); // ANTI_ALIAS_FLAG DITHER_FLAG
        clearPaint.setColor(0xffffffff);
        updateBrush();
        //
    }
    public void changeTask(Task t) {
        task = t;
        needRedraw = true;
        invalidate();
    }
    public void onSizeChanged(int w, int h, int oldw, int oldh) {
        try {
            super.onSizeChanged(w, h, oldw, oldh);
            App.getInstance().putInt(c.gamew, w);
            App.getInstance().putInt(c.gameh, h);
            bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888);
            canvas = new Canvas(bitmap);
            rect.set(0.0f, 0.0f, (float) w, (float) h);
        } catch (Exception e2) {
            e2.printStackTrace();
        }
    }
    public void undo(){
        currentPathId--;
        needRedraw = true;invalidate();
    }
    public void redo(){
        currentPathId++;
        needRedraw = true;invalidate();
    }
    public void updateTips(){
        task.makeTips();
        needRedraw = true;invalidate();
    }
    public void updateBrush(){
        brushManager.updatePaint(paint);
    }

    public void redraw(){
        needRedraw = false;
        canvas.drawRect(rect, clearPaint);
        task.draw(canvas);
        for (int i=0; i<currentPathId; i++){
            canvas.drawPath(paths.get(i), paint);
        }
        pathsListener.updatePaths(paths.size(), currentPathId);
    }
    @Override
    public void onDraw(Canvas canvasGlobal) {
        super.onDraw(canvas);
        try {
            if (needRedraw)
                redraw();
            canvas.drawPath(path, paint);
            canvasGlobal.drawBitmap(this.bitmap, (Rect) null, this.rect, paintBmp);
        } catch (Exception exc) {
            exc.printStackTrace();
        }
    }
    public boolean onTouchEvent(MotionEvent motionEvent) {
        float x = motionEvent.getX();
        float y = motionEvent.getY();
        int action = motionEvent.getAction();
        if (action == 0) { // ACTION_DOWN pressed gesture has started, the motion contains the initial starting location.
            for (int i=paths.size()-1;i>=currentPathId; i--)
                paths.remove(i);
            path.reset();
            sx = x;
            sy = y;
            paint.setStrokeWidth((float) penSize);
            brushManager.updatePaint(paint);
            path.moveTo(x, y);

        } else if (action == 1) { // ACTION_UP
            path.lineTo(sx, sy);
            canvas.drawPath(path, paint);
            paths.add(path);
            path = new Path();
            pathsListener.updatePaths(paths.size(), ++currentPathId);
        } else if (action != 2) {
            return false;
        } else { /// 2 = ACTION_MOVE
            if (Math.abs(x - sx) >= 0.0f || Math.abs(y - sy) >= 0.0f) {
                path.quadTo(sx, sy, (sx + x) / 2.0f, (sy + y) / 2.0f);
                sx = x;
                sy = y;
                task.updateProgress((int)x, (int)y);
                progressListener.updateProgress(task.getProgress());
            }
        }
        invalidate();
        return true;
    }
}

File: app/src/main/java/artwork/kotlin/GameView.kt
package artwork.kotlin

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.MotionEvent
import androidx.appcompat.widget.AppCompatImageView
import java.util.*
import artwork.java.App
import artwork.java.c

class GameView : AppCompatImageView {
    lateinit var canvas: Canvas
    var rect: RectF = RectF()
    lateinit var bitmap: Bitmap
    var bmpPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    var paint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    var clearPaint: Paint = Paint(Paint.ANTI_ALIAS_FLAG)
    var path: Path = Path()
    var sx: Float = 0f
    var sy: Float = 0f
    var penSize: Int = 10
    var paths: Vector<Path> = Vector()
    var currentPathId: Int = 0
    var needRedraw: Boolean = true
    lateinit var progressListener: ProgressListener
    lateinit var pathsListener: UpdatePathsListener
    var brushManager: BrushManager = BrushManager()
    var task: Task? = null
    /**
     * @param context the context
     */
    constructor(context: Context) : super(context)
    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs) {
        progressListener = context as ProgressListener
        pathsListener = context as UpdatePathsListener
        paint.isAntiAlias = true
        paint.style = Paint.Style.STROKE
        paint.strokeJoin = Paint.Join.ROUND
        paint.strokeCap = Paint.Cap.ROUND
        paint.setPathEffect(CornerPathEffect(10.0f))
        clearPaint.color = 0xffffffff.toInt()
        updateBrush()
    }
    fun changeTask(t: Task) {
        task = t
        needRedraw = true
        invalidate()
    }
    public override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        try {
            super.onSizeChanged(w, h, oldw, oldh)
            App.getInstance().putInt(c.gamew, w)
            App.getInstance().putInt(c.gameh, h)
            bitmap = Bitmap.createBitmap(w, h, Bitmap.Config.ARGB_8888)
            canvas = Canvas(bitmap)
            rect[0.0f, 0.0f, w.toFloat()] = h.toFloat() // ???
        } catch (e2: Exception) {
            e2.printStackTrace()
        }
    }

    fun undo() {
        if (0 != currentPathId) {
            currentPathId--
            needRedraw = true
            invalidate()
        }
    }

    fun redo() {
        if (currentPathId != paths.size) {
            currentPathId++
            needRedraw = true
            invalidate()
        }
    }

    fun updateTips() {
        task?.let{
            it.makeTips()
            needRedraw = true
            invalidate()
        }
    }
    fun updateBrush() {
        brushManager.updatePaint(paint)
    }

    fun redraw() {
        needRedraw = false
        canvas.drawRect(rect, clearPaint)
        task?.draw(canvas)
        for (i in 0 until currentPathId) {
            canvas.drawPath(paths[i], paint)
        }
        pathsListener.updatePaths(paths.size, currentPathId)
    }
    public override fun onDraw(canvasGlobal: Canvas) {
        super.onDraw(canvas)
        try {
            if (needRedraw) redraw()
            canvas.drawPath(path, paint)
            canvasGlobal.drawBitmap(bitmap, null as Rect?, this.rect, bmpPaint)
        } catch (exc: Exception) {
            exc.printStackTrace()
        }
    }
    override fun onTouchEvent(motionEvent: MotionEvent): Boolean {
        val x = motionEvent.x
        val y = motionEvent.y
        val action = motionEvent.action
        if (action == 0) { // ACTION_DOWN pressed gesture has started, the motion contains the initial starting location.
            for (i in paths.size - 1 downTo currentPathId) paths.removeAt(i)
            path.reset()
            sx = x
            sy = y
            paint.strokeWidth = penSize.toFloat()
            brushManager.updatePaint(paint)
            path.moveTo(x, y)
        } else if (action == 1) { // ACTION_UP
            path.lineTo(sx, sy)
            canvas.drawPath(path, paint)
            paths.add(path)
            path = Path()
            pathsListener.updatePaths(paths.size, ++currentPathId)
        } else if (action != 2) {
            return false
        } else { /// 2 = ACTION_MOVE
            if (x!=sx && y!=sy) {
                path.quadTo(sx, sy, (sx + x) / 2.0f, (sy + y) / 2.0f)
                sx = x
                sy = y
                task?.updateProgress(x.toInt(), y.toInt())
                task?.let { progressListener.updateProgress(it.progress) }
            }
        }
        invalidate()
        return true
    }
}
